### ISSUE TITLE
Python REPL

### ISSUE THREAD

**Issue by piranna** [atsepkov/RapydScript#109]
Since RapydScript allow to automatically exec the just compiled script, and has a syntax (almost) identical to regular Python, it would be trivial to have Python REPL in Node.js. It would be mostly reuse and exec some components of the current CLI by default (like the `execute` flag) and add a `python` entry on the `bin` section of `package.json`. This could be useful both for NodeOS to have a Python interpreter but also to node-gyp to don't depend on the Python binary but instead use a Javascript based one.

**Comment by atsepkov** [comment]
I don't completely understand what you're asking. RapydScript already comes with a REPL, but I don't understand what you're trying to do with Python. What would the two projects you mention do with the REPL?

**Comment by piranna** [comment]
node-gyp are some bindings for gyp, that's used to compile Node js binary modules. gyp is based on Python, so by using RapydScript as interpreter we would not need it and use Node.js instead. Regarding NodeOS, it's an operating system build on top of Linux and using Node.js for the user space, so with RapydScript we could run Python scripts on it... including the ones from gyp :-)

**Comment by atsepkov** [comment]
It's not as simple as it seems. While the syntax is very similar to that of Python, the underlying fundamentals of the language are still those of JavaScript. "one" + 2 is a TypeError in python, but in JavaScript it's "one2". Operator overloading is allowed in Python, but not in JavaScript. These may seem like minor differences, but they're very significant for lower level libraries that rely on them, and you can't advertise something as a Python compiler with limitations like these.

There are already some dubious issues reported claiming bugs in RapydScript because the functionality doesn't match that of Python. If we claimed to be Python, these reports would be much more common and indeed would be classified as bugs. It would be a big headache.

**Comment by piranna** [comment]
I see, it's mostly a Javascript interpretar with a Python syntax... :-/ I think this issues would be addressed and make a real Python transpiler, but I admit that maybe it's not the long term focus of RapydScript and should be a different project.

I have been thinking about building a high performant Python-2-Javascript transpiler since a long time, and I think how RapydScript works is the correct approach to do it, and by using ES& features we can get almost a 1:1 relationship since all the ES6 goodies are borrowed from Python :-P The remaining parts would be implemented with a minimal runtime, for example I've already seen libraries that implement Python-like operator overloading and multiple inheritance in Javascript, so it would only be a matter of use them to fill the ES6 missing parts :-)

**Comment by atsepkov** [comment]
I was one of the maintainers for Pyjamas project, which was a "real" Python transpiler. I've seen the good, the bad, and the ugly. RapydScript is a result of the issues I've experienced dealing with a full transpiler (http://blog.pyjeon.com/2012/07/29/why-pyjamas-isnt-a-good-framework-for-web-apps/), it's based on a revelation that the goal of a transpiler is to minimize context switch, and that I didn't need 100% of Python in the browser world. As a result, we now have 90% of Python functionality, with no overhead - that's the Pareto principle at work.

You're right that we now have ES6, which borrows heavily from Python. Back in the days of Pyjamas, we did not have that luxury. However, the fundamental differences (lax error handling, treatment of null values and arguments) will not change. You will not have "real" Python until you address those issues, and by the time you do, your "language" will perform at a fraction of speed of existing RapydScript. That may be fine for "educational" purposes, but no one will adopt your language in the real world (see Brython and Skulpt as examples of projects that failed here). If you do want to build a more pure Python language, feel free to fork. RapydScript is a good starting point, and you'll get quite far by simply wrapping primitive JS types in Pythonic classes. But I'm not going down that road again myself, I've been there and I can tell you that performance will kill you.

I'm closing this issue for now, let me know if you want me to reopen it.
